New algorithm to refine lattice parameters across runs with separate U matrix/orientation

RichardWaiteSTFC opened the pull request:
**Description of work.**

New algorithm `FindGlobalBMatrix` to refine the lattice parameters (encoded in the B matrix) across many runs with separate U matrix/orientation which is useful for when the goniometer is inaccurate/not well known.

The quantity minimised is the average of the square of the difference in `QSample` between peak the peak at integer HKL (the minimisation is performed by scipy `leastsq` function). The average rather than the sum of the squared residuals is used so as not to penalise the indexing of more peaks as the UB matrix becomes more accurate.

`FindGlobalBMatrix` adds a different UB to each peak workspace that can be indexed (the algorithm ensures consistent indexing across runs - e.g. not swapped `H<->K` and `L->-L`) - however all the workspaces have the same lattice parameters (it's just U that's different between runs)

**To test:**

Use some SXD data of cubic NaCl (a=b=c=6.65, alpha=beta=gamma = 90) and see how close the refined parameters are to nominal

(1) Run this script to load the data and find peaks
```
runs = list(range(32864, 32866, 1))

def process_run(runno):
    w = Load(Filename='SXD' + str(runno) +'.raw', OutputWorkspace='SXD' + str(runno))
    LoadParameterFile(Workspace='SXD' + str(runno), 
        Filename=r'\\olympic\Babylon5\Public\RWaite\SXD32863_calibration.xml') # updated calibration
    NormaliseByCurrent(InputWorkspace='SXD' + str(runno), OutputWorkspace='SXD' + str(runno))
    CropWorkspace(InputWorkspace='SXD' + str(runno), OutputWorkspace='SXD' + str(runno), XMin=1000, XMax=10000)
    # SetGonio doesn't recognise wccr log so have to manually average 
    wccr = w.run().getPropertyAsSingleValueWithTimeAveragedMean('wccr')
    SetGoniometer(Workspace='SXD' + str(runno), Axis0=str(wccr)+',0,1,0,1') 
    # Find peaks
    FindSXPeaks(InputWorkspace='SXD' + str(runno), PeakFindingStrategy='AllPeaks', AbsoluteBackground=5,
        ResolutionStrategy='AbsoluteResolution', XResolution=400, PhiResolution=3, TwoThetaResolution=3, 
        OutputWorkspace='SXD' + str(runno) +'_peaks')
    FilterPeaks(InputWorkspace='SXD' + str(runno) +'_peaks', OutputWorkspace='SXD' + str(runno) +'_peaks', 
        Criterion='!=', BankName='bank1')  # remove peaks in Bank 1 as calibration poor

for irun, runno in enumerate(runs):
    process_run(runno)
    if irun == 0:
        AllPeaks = CloneWorkspace(InputWorkspace='SXD' + str(runno) + '_peaks')
    else:
        AllPeaks = CombinePeaksWorkspaces(AllPeaks, 'SXD' + str(runno) + '_peaks',  CombineMatchingPeaks=False)
```
 (2) Run this to refine the global B matrix (it is slowed down by the call to `FindUBUsingLatticeParameters` to find the initial UB)
```
FindGlobalBMatrix(PeakWorkspaces = ['SXD32864_peaks', 'SXD32865_peaks'],
    a=5.4, b=5.8,c=5.9,alpha=88,beta=88,gamma=89)
```
It should print this in the log  - format is  [ a, b, c, alpha, beta, gamma]
```
Lattice Parameters: [ 5.653631  5.65586   5.656385 89.99074  90.077986 90.003773]
Parameter Errors  : [0.007926 0.011577 0.013786 0.181732 0.176361 0.132551]
```
(3) Play around with the initial parameters - it should return the same result (typically we know the rough lattice parameter within 0.1 Ang so already the initial guess used above is quite an extreme example)

(4) Compare this to running `FindUBUsingLatticeParameters` with a much better guess on the combined workspaces which has a single orientation for all runs that is rotated by the goniometer matrix.
```
FindUBUsingLatticeParameters(PeaksWorkspace='AllPeaks', a=5.65, b=5.65, c=5.65, alpha=90, beta=90, gamma=90)
```
It should print 
```
Lattice Parameters:    5.641454    5.697618    5.670424   89.542192   89.929136   90.329593   182.254616
Parameter Errors  :    0.014173    0.017662    0.008407    0.196720    0.167132    0.228287    0.775815
```
You will see the lattice parameters are less consistent with cubic NaCl lattice. 

<!-- Instructions for testing. -->

Fixes #31909 


<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

RichardWaiteSTFC committed:
Add new algorithm to refine global B

RichardWaiteSTFC committed:
Add functional unit tests for FindGlobalBMatrix

No tests yet for input validation or edge cases etc.

RichardWaiteSTFC committed:
Make use of child algorithm calls

RichardWaiteSTFC committed:
Add tolerance argument for indexing

RichardWaiteSTFC committed:
Fix bug in transfrom passed to TransfromHKL

RichardWaiteSTFC committed:
Add tests for input validation and refactor to avoid duplicate code

RichardWaiteSTFC committed:
Add assertions for U matrix to functional tests

RichardWaiteSTFC committed:
Refactor to take initial indexing outside main exec

RichardWaiteSTFC committed:
Use UB that indexes most peaks as reference

RichardWaiteSTFC committed:
Transform existing UB if exists and add test for transform

RichardWaiteSTFC committed:
Refactor to use function to generate peaks in tests

RichardWaiteSTFC committed:
Add errors on lattice parameters

Would normally scale covariance matrix by chisq if minimised square of residuals. Here we minimise the average square of residuals (so penalty for indexing peaks). However, I think it's still correct to normalise by dof and the magnitude of the errors agree well with other methods e.g. FindUBUsingLatticeParameters.

RichardWaiteSTFC committed:
Assert B matrix equal in tests

RichardWaiteSTFC committed:
Add algorithm documentation and release notes

RichardWaiteSTFC committed:
Merge branch 'master' into 31909_new_algorithm_to_refine_global_B

RichardWaiteSTFC committed:
Correct naming of doc rst file

DannyHindson commented:
I just had a look at this PR to see if it was ready for review. It looks like a CodeConventions system test has found a genuine problem with the parameter names on the new algorithm. It doesn't like the parameters starting with a lower case letter. 
So you'll either need to change the names or if you want to maintain consistency with other algorithms (eg CalculateUMatrix) then you'll need to add the new algorithm as an exclusion into CodeConventions.py

RichardWaiteSTFC committed:
Add lowercase args to code conventions

RichardWaiteSTFC commented:
> I just had a look at this PR to see if it was ready for review. It looks like a CodeConventions system test has found a genuine problem with the parameter names on the new algorithm. It doesn't like the parameters starting with a lower case letter.
> So you'll either need to change the names or if you want to maintain consistency with other algorithms (eg CalculateUMatrix, FindUBUsingLatticeParameters) then you'll need to add the new algorithm as an exclusion into CodeConventions.py

Thanks for spotting this, have gone with the option of adding the lowercase args to `CodeConventions.py` for consistency with the other algorithms.

RichardWaiteSTFC commented on Framework/PythonInterface/test/python/plugins/algorithms/FindGlobalBMatrixTest.py:
```diff
@@ -0,0 +1,150 @@
+        self.assert_matrix([peaks1, peaks2], np.eye(3), getUMatrix, delta=5e-2)
+
+    def test_handles_inaccurate_goniometer(self):
+        # create two peak tables with UB corresponding to different lattice constant, a
```
On a different branch but need to remove this comment- it applies to  the test above

RichardWaiteSTFC committed:
Clarify and correct typos in some comments

DannyHindson reviewed (changes requested):
Have tested this out and works nicely. Good job in quite a short space of time!

I've added a few v minor comments.

Also I found that by picking daft starting parameters I was able to make the leastsq optimisation fall over when it tries negative values for a, b or c in the call to CalculateUmatrix eg with:
```
FindGlobalBMatrix(PeakWorkspaces = ['SXD32864_peaks', 'SXD32865_peaks'],
    a=0.5, b=5.8,c=5.9,alpha=88,beta=88,gamma=89)
```
Could this error situation be caught? It gives quite a verbose\technical error at the moment:

![image](https://user-images.githubusercontent.com/56295817/125969334-4a09807d-6595-40be-b0c1-408ca85549d4.png)

RichardWaiteSTFC committed:
Add try/except to catch when CalculateUMatrix fails

RichardWaiteSTFC commented:
I can't actually reproduce the error you saw, but I put the call to calculateUMatrix in a try (which I thought was quicker than having an if statement execute every time in a fit iteration). Can you confirm that fixes it?

DannyHindson commented:
> I can't actually reproduce the error you saw, but I put the call to calculateUMatrix in a try (which I thought was quicker than having an if statement execute every time in a fit iteration). Can you confirm that fixes it?

As well as setting a=0.5 I had also set the Tolerance to a large value (eg 10.0). With this "a" and Tolerance setting I get the error I pasted in earlier. If I remove the tolerance parameter or set it explicitly back to the default value of 0.15, the next call to FindGlobalBMatrix still fails with the same error. So it seems a bit sticky. I'll have a look at the change you've made though thanks

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/FindGlobalBMatrix.py:
```diff
@@ -0,0 +1,263 @@
+        if success:
+            # calculate errors
+            dof = sum([self.child_IndexPeaks(ws, RoundHKLs=True) for ws in ws_list]) - len(alatt0)
+            err = np.sqrt(abs(np.diag(cov)) * (info['fvec'] ** 2).sum() / dof)
```
Again by putting daft parameters into the FindGlobalBMatrix call I occasionally get an error from this line because "cov" is set to None:

![image](https://user-images.githubusercontent.com/56295817/126135312-b3f47146-ac3f-425d-9ea0-a4434984f4fd.png)

I think the code should check here if cov is not None before trying to calculate err.

In case it helps, the other return parameters from the call to leastsq in this situation are:
ier=2
msg = 'The relative error between two consecutive iterates is at most 0.000000'
alatt and info are populated with what looks like sensible data

RichardWaiteSTFC replied:
Have added this check and a comment to explain that the cov matrix is set to None is singular (but fit can still be a success)

DannyHindson reviewed (changes requested):
This can wait until you're back but here are a few other suggestions around improving the error handling in the edge cases where the optimisation fails

RichardWaiteSTFC committed:
Move try put of obj func, check cov not none, remove duplciate self

Try now around the leastsq func as suggested - it is also now correctly excepts ValueError

RichardWaiteSTFC commented on Framework/PythonInterface/plugins/algorithms/FindGlobalBMatrix.py:
```diff
@@ -0,0 +1,263 @@
+            if nindexed >= _MIN_NUM_INDEXED_PEAKS:
+                try:
+                    self.child_CalculateUMatrix(wsname, *x0)
+                except RuntimeError:
```
Yes it was due to not returning an array - I have done as suggested and put the try around the call to leastsq as it's also quicker than having in the objective function (I think).

RichardWaiteSTFC replied:
FYI I can reproduce the `ValueError `with the line
```
FindGlobalBMatrix(PeakWorkspaces = ['SXD32864_peaks', 'SXD32865_peaks'],
    a=0.01, b=5.8,c=5.9,alpha=88,beta=88,gamma=89, Tolerance = 10)
```

RichardWaiteSTFC commented:
Just an update - I'd like to use `scipy.least_squares` minimiser that supports bounds on parameters but it isn't supported for scipy < v1.1.0  - so it wouldn't work for RHEL7 atm.

DannyHindson reviewed (approved):
This looks good now. Have retested and the errors from bad input parameters are caught\handled nicely
